### PR TITLE
feat(atomic): add segmented scrollable component basic structure

### PR DESCRIPTION
**Pull request by mbenkirane-coveo**
Update:

Screen Shot 2022-06-17 at 1 00 40 PM

Prev PR draft issues:
Trying to get the nested elements from the shadow dom returned undefined, so I tried to get the nested elements using the regular dom. I was able to get the elements using the tag name, then mapping them on the atomic-segmented-facet components. Styling wasn't applied properly, more specifically, the "overflow-x-scroll" doesn't work. I also need to remove the mapped duplicated components.

Been looking at this [StackOverflow](https://stackoverflow.com/questions/52421298/web-components-how-to-work-with-children) link, but didn't work for me.

Blockers:
- Child nodes in shadow dom returns undefined
- When using regular dom, styling not applied to mapped components
- Need to remove duplicated children as the removeChild() doesn't seem to work

### CODE REVIEW

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+  connectedCallback() {
+    this.children = Array.from(this.host.children);
+    console.log(this.children);
+  }
+
+  render() {
+    return (
+      <div class="flex overflow-x-scroll">
+        <slot></slot>
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:19]: https://stenciljs.com/docs/templating-jsx#slots

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  tag: 'atomic-segmented-facet-scrollable',
+  shadow: true,
+})
+export class AtomicSegmentedFacetScrollable {
+  @Element() private host!: HTMLElement;
+  @State() children: Array<any> = [];
+
+  connectedCallback() {
+    this.children = Array.from(this.host.children);
+    console.log(this.children);
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:13]: Screen Shot 2022-06-15 at 1 20 37 PM

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import {Component, Element, h, State} from '@stencil/core';
+
+@Component({
+  tag: 'atomic-segmented-facet-scrollable',
+  shadow: true,
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:5]: since there will be styles, it should have a shadow DOM

File: packages/atomic/src/pages/index.html [hunk 4]
```diff
@@ -194,6 +194,9 @@
           </atomic-facet-manager>
         </atomic-layout-section>
         <atomic-layout-section section="main">
+          <atomic-segmented-facet-scrollable>
```
Comment (ThibodeauJF) [inline packages/atomic/src/pages/index.html:197]: Will be easier to test if you have an example

File: packages/atomic/src/components/facets/atomic-segmented-facet/atomic-segmented-facet.tsx [hunk 5]
```diff
@@ -146,11 +146,9 @@ export class AtomicSegmentedFacet
...
-    return (
-      <div class="flex">
-        <b class="inline-block my-3 mr-2">{this.label}</b>
-        {this.renderValues()}
-      </div>
-    );
+    return [
+      <b class="inline-block my-3 mr-2">{this.label}</b>,
+      this.renderValues(),
+    ];
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/facets/atomic-segmented-facet/atomic-segmented-facet.tsx:152]: Why this change?

Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet/atomic-segmented-facet.tsx:152]: Still haven't merged in the newest branch

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+})
+export class AtomicSegmentedFacetScrollable {
+  @Element() private host!: HTMLElement;
+  @State() children: Array<any> = [];
+
+  connectedCallback() {
```
[the remark this answers is not in the document]
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:11]: Only difference is that it rerenders when a component is reattached? In what case would this component get detached and reattached?

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    this.children = Array.from(this.host.children);
+    console.log(this.children);
+  }
+
+  render() {
+    return (
+      <div class="flex overflow-x-scroll">
```
[the remark this answers is not in the document]
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:18]: For now, adding a width doesn't make the search bar appear. I'm still trying to figure out why.

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx (hunk 1 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:19]: Much simpler than my implementation

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:18]: Seems like style isn't being applied to the shadow root. tried to style the shadow roots inner html and for some reason a new div gets created thats styled but the slots inside don't contain the segmented facets

Screen Shot 2022-06-15 at 5 52 10 PM

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (ThibodeauJF) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:11]: We generally use it for all components. Facets for example get moved around by us, but we never know what customers will do with our custom elements, they might have a use case for moving them around.

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    //     : this.host.shadowRoot;
+
+    console.log(this.host.shadowRoot);
+  }
+
+  render() {
+    return (
+      <div class="flex">
+        <div class="border border-neutral bg-background text-on-background hover:bg-neutral-light focus-visible:bg-neutral-light no-outline rounded">
+          <atomic-icon
```
Comment (mbenkirane-coveo) [inline packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:30]: When adding a label and especially a label with 2 or more words the arrow buttons aspect ratio changes. I need to find a way to keep it constant and also adjust the size of the arrow icon as well.

Comment (ThibodeauJF) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:30]: The segmented facet have a height of 35px without a label but of 41px with a label. We should set a static height on the facet and use flex to align the items vertically.

After that it should be easy to use a good scrollable aspect ratio

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    // this.children = Array.from(this.host.children);
+    // console.log(this.children);
+
+    // const shadowRoot =
+    //   this.host.shadowRoot === null
+    //     ? this.host.attachShadow({mode: 'open'})
+    //     : this.host.shadowRoot;
+
+    console.log(this.host.shadowRoot);
+  }
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:24]: won't need his in the end I guess? same for the children state

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx (hunk 8 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:30]: I fixed some of the issues ! I will implement your solution as well. Thanks !

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx (hunk 9 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:24]: yup, I'll remove it

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+})
+export class AtomicSegmentedFacetScrollable {
+  @Element() private host!: HTMLElement;
+
+  render() {
+    return (
+      <div class="flex h-9">
+        <div class="flex shrink-0 basis-8 justify-center items-center border border-neutral bg-background text-on-background hover:bg-neutral-light focus-visible:bg-neutral-light no-outline rounded">
+          <atomic-icon class="w-3.5" icon={ArrowLeftIcon}></atomic-icon>
+        </div>
```
Comment (btaillon) [inline packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:18]: I would avoid having clickable divs, since that's bad for accessibility. I would change them into `button` elements.

Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:18]: Sounds good!

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import ArrowLeftIcon from 'coveo-styleguide/resources/icons/svg/arrow-left-rounded.svg';
+import {Button} from '../../common/button';
+
+@Component({
+  tag: 'atomic-segmented-facet-scrollable',
+  styleUrl: 'atomic-segmented-facet-scrollable.pcss',
+  shadow: true,
+})
+export class AtomicSegmentedFacetScrollable {
+  @Element() private host!: HTMLElement;
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:12]: unused
```suggestion
```

Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:12]: it will be used to add functionality for the arrows

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  render() {
+    return (
+      <div class="flex h-9">
+        <Button
+          style="square-neutral"
+          class="flex shrink-0 basis-8 justify-center items-center rounded"
+          ariaHidden="true"
+        >
+          <atomic-icon class="w-3.5" icon={ArrowLeftIcon}></atomic-icon>
+        </Button>
```
[the remark this answers is not in the document]
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:23]: sounds good

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        <div class="wrapper-segmented flex flex-row overflow-x-scroll">
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:24]: detail, doesn't really matter since you hide the scrollbar already
```suggestion
        <div class="wrapper-segmented flex flex-row overflow-x-auto">
```

**Review by ThibodeauJF — APPROVED** [review]
💯

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx (hunk 13 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:24]: Messes with the overflow.
Screen Shot 2022-06-21 at 12 10 12 PM

File: packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/facets/atomic-segmented-facet-scrollable/atomic-segmented-facet-scrollable.tsx:23]: will do this on this [PR](https://github.com/coveo/ui-kit/pull/2130)
